controllers/imagerepository_controller.go

```diff
@@ -577,32 +590,216 @@ func (r *ImageRepositoryReconciler) EnsureSecret(ctx context.Context, imageRepos
+				log.Error(err, "Failed to link secret to service account", "SecretName", secretName, l.Action, l.ActionUpdate)
 				return err
 			}
-			serviceAccount.Secrets = append(serviceAccount.Secrets, corev1.ObjectReference{Name: secretName})
-			serviceAccount.ImagePullSecrets = append(serviceAccount.ImagePullSecrets, corev1.LocalObjectReference{Name: secretName})
-			if err := r.Client.Update(ctx, serviceAccount); err != nil {
-				log.Error(err, "failed to update service account", l.Action, l.ActionUpdate)
-				return err
+		}
+	}
+	return nil
+}
+
+// linkSecretToServiceAccount ensures that the given secret is linked with the provided service account.
+func (r *ImageRepositoryReconciler) linkSecretToServiceAccount(ctx context.Context, secretNameToAdd, namespace string) error {
+	log := ctrllog.FromContext(ctx)
+
+	serviceAccount := &corev1.ServiceAccount{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: buildPipelineServiceAccountName, Namespace: namespace}, serviceAccount)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, "Failed to read pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionView)
+		return err
+	}
+
+	// check if secret is already linked and add it only if it isn't to avoid duplication
+	secretLinked := false
+	for _, serviceAccountSecret := range serviceAccount.Secrets {
+		if serviceAccountSecret.Name == secretNameToAdd {
+			secretLinked = true
+			break
+		}
+	}
+	if !secretLinked {
+		// add it only to Secrets and not in imagePullSecrets which is used only for the image pod itself needs to run
+		serviceAccount.Secrets = append(serviceAccount.Secrets, corev1.ObjectReference{Name: secretNameToAdd})
+
+		if err := r.Client.Update(ctx, serviceAccount); err != nil {
+			log.Error(err, "Failed to update pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionUpdate)
+			return err
+		}
+		log.Info("Added secret link to pipeline service account", "SecretName", secretNameToAdd, "ServiceAccountName", buildPipelineServiceAccountName, l.Action, l.ActionUpdate)
+	}
+	return nil
+}
+
+// unlinkSecretFromServiceAccount ensures that the given secret is not linked with the provided service account.
+func (r *ImageRepositoryReconciler) unlinkSecretFromServiceAccount(ctx context.Context, secretNameToRemove, namespace string) error {
+	log := ctrllog.FromContext(ctx)
+
+	serviceAccount := &corev1.ServiceAccount{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: buildPipelineServiceAccountName, Namespace: namespace}, serviceAccount)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, "Failed to read pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionView)
+		return err
+	}
+
+	isSecretUnlinked := false
```

Comment:
The name is confusing.
`if isSecretUnlinked { update` but when it's already unlinked no need to update...

Reply:
renamed to unlinkSecret